algo/src/main/java/org/neo4j/graphalgo/impl/ParallelUnionFindQueue.java

```diff
@@ -61,43 +64,64 @@
     public ParallelUnionFindQueue(Graph graph, ExecutorService executor, int minBatchSize, int concurrency) {
         super(graph);
         this.executor = executor;
-        nodeCount = Math.toIntExact(graph.nodeCount());
+        this.nodeCount = Math.toIntExact(graph.nodeCount());
         this.batchSize = ParallelUtil.adjustBatchSize(nodeCount, concurrency, minBatchSize);
-        stepSize = ParallelUtil.threadSize(batchSize, nodeCount);
     }
 
     @Override
     public DisjointSetStruct compute() {
-        final List<Future<?>> futures = new ArrayList<>(stepSize);
+        int stepSize = ParallelUtil.threadSize(batchSize, nodeCount);
+        final List<Future<?>> futures = new ArrayList<>(2 * stepSize);
         final BlockingQueue<DisjointSetStruct> queue = new ArrayBlockingQueue<>(stepSize);
+        AtomicInteger expectedStructs = new AtomicInteger();
 
-        Phaser phaser = new Phaser();
-        int steps = 0;
         for (int i = 0; i < nodeCount; i += batchSize) {
-            futures.add(executor.submit(new UnionFindTask(queue, i, phaser)));
-            ++steps;
+            futures.add(executor.submit(new UnionFindTask(queue, i, expectedStructs)));
         }
-        phaser.awaitAdvance(phaser.getPhase());
+        int steps = futures.size();
 
         for (int i = 1; i < steps; ++i) {
-            futures.add(executor.submit(() -> {
-                try {
-                    final DisjointSetStruct a = queue.take();
-                    final DisjointSetStruct b = queue.take();
-                    queue.add(a.merge(b));
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    throw new RuntimeException(e);
-                }
-            }));
+            futures.add(executor.submit(() -> mergeTask(queue, expectedStructs, DisjointSetStruct::merge)));
         }
 
-        await(futures);
+        awaitTermination(futures);
         return getStruct(queue);
     }
 
-    private void await(final List<Future<?>> futures) {
-        ParallelUtil.awaitTermination(futures);
+    static <T> void mergeTask(
```

Comment:
perhaps we can generalize these in another PR later,

i.e. the expected CAS
and the actual merge operation with the push check
and move it out to a utility or (superclass minus inlining costs)